File: Challenge-002-Preparing-Delimited-Data/Philip Mannering/Challenge-002-Preparing-Delimited-Data.py
"""
Challenge 002 - Preparing Delimited Data
Author: Philip Mannering
Date: 2018-09-07
"""

import pandas as pd

df = pd.read_csv('../files/input.csv')
# Read in the answer and convert the 3rd column to datetime
output = pd.read_csv('../files/output.csv', parse_dates=[2])

#%% Parse

# text to columns
df = df.Field_1.str.split(',', expand=True)
df.columns = ['Poem','Poem_ID','Poem_Read_Date']

# Remove quotes and set datatype
df.Poem = df.Poem.str.strip('"')
df.Poem_ID = df.Poem_ID.astype('int64')
df.Poem_Read_Date = pd.to_datetime(df.Poem_Read_Date)


#%% Check answer

# Datatypes must match
df.equals(output)

File: Challenge-002-Preparing-Delimited-Data/P-Man/Challenge-002-Preparing-Delimited-Data.R
# Challenge 002 - Preparing Delimited Data
# Author: Philip Mannering
# Date: 2018-09-07

# All the glorious regex functions!
library(stringr)

setwd('~/Github/ARP/Challenge-002-Preparing-Delimited-Data/Philip Mannering/')

# Read in input
df <- read.csv('../files/input.csv')

# Split field by a comma
splitted <- str_split_fixed(df$Field_1,',',3)

# Remove single and double quotes
splitted <- apply(splitted, 2, function(x) str_remove_all(x,'[\'"]'))

# Change matrix to dataframe
splitted <- as.data.frame(splitted)
names(splitted) = c('Poem','Poem_ID','Poem_Read_Date')

# Change datatype
ans <- transform(splitted,  Poem = as.character(Poem),
                            Poem_ID = as.numeric(as.character(Poem_ID)),
                            Poem_Read_Date = as.Date(Poem_Read_Date, '%d-%b-%y'))


# Check the output
op <- read.csv('../files/output.csv', as.is=T,
               colClasses = c('character','numeric','Date'))
identical(ans,op)

